lang/csharp/src/apache/test/Generic/GenericTests.cs

```diff
@@ -37,6 +37,45 @@ private static void test<T>(string s, T value)
             Assert.AreEqual(value, output);
         }
 
+        [Test]
```

Comment:
We need more unit tests here. Something like this:

```
        private static IEnumerable<TestCaseData> ConvertsDefaultToLogicalTypeSource = new List<TestCaseData>()
        {
            new TestCaseData(@"{""type"": ""bytes"", ""logicalType"": ""decimal"", ""precision"": 4}", @"""0""", (decimal)0),
            new TestCaseData(@"{""type"": ""bytes"", ""logicalType"": ""decimal"", ""precision"": 4}", @"""1234""", (decimal)1234),
            new TestCaseData(@"{""type"": ""bytes"", ""logicalType"": ""decimal"", ""precision"": 4}", @"""1.234""", (decimal)1.234),
            new TestCaseData(@"{""type"": ""string"", ""logicalType"": ""uuid""}", @"""00000000-0000-0000-0000-000000000000""", new Guid()),
            new TestCaseData(@"{""type"": ""string"", ""logicalType"": ""uuid""}", @"""00000000000000000000000000000000""", new Guid()),
            new TestCaseData(@"{""type"": ""string"", ""logicalType"": ""uuid""}", @"""12345678-1234-5678-1234-123456789012""", new Guid("12345678-1234-5678-1234-123456789012")),
            new TestCaseData(@"{""type"": ""string"", ""logicalType"": ""uuid""}", @"""12345678123456781234123456789012""", new Guid("12345678-1234-5678-1234-123456789012")),
            new TestCaseData(@"{""type"": ""int"", ""logicalType"": ""date""}", "0", DateTime.UnixEpoch),
            new TestCaseData(@"{""type"": ""int"", ""logicalType"": ""date""}", "123456", DateTime.UnixEpoch.AddDays(123456)),
            new TestCaseData(@"{""type"": ""long"", ""logicalType"": ""time-micros""}", "0", new TimeSpan()),
            new TestCaseData(@"{""type"": ""long"", ""logicalType"": ""time-micros""}", "123456", new TimeSpan(123456*TimeSpan.TicksPerMillisecond/1000)),
            new TestCaseData(@"{""type"": ""int"", ""logicalType"": ""time-millis""}", "0", new TimeSpan()),
            new TestCaseData(@"{""type"": ""int"", ""logicalType"": ""time-millis""}", "123456", new TimeSpan(0, 0, 0, 0, 123456)),
            new TestCaseData(@"{""type"": ""long"", ""logicalType"": ""timestamp-micros""}", "0", DateTime.UnixEpoch),
            new TestCaseData(@"{""type"": ""long"", ""logicalType"": ""timestamp-micros""}", "123456", DateTime.UnixEpoch.AddTicks(123456*TimeSpan.TicksPerMillisecond/1000)),
            new TestCaseData(@"{""type"": ""long"", ""logicalType"": ""timestamp-millis""}", "0", DateTime.UnixEpoch),
            new TestCaseData(@"{""type"": ""long"", ""logicalType"": ""timestamp-millis""}", "123456", DateTime.UnixEpoch.AddMilliseconds(123456))
        };

        [TestCaseSource(nameof(ConvertsDefaultToLogicalTypeSource))]
        public void ConvertsDefaultToLogicalType(string typeDefinition, string defaultDefinition, object expected)
        {
            var writerSchemaString = @"{
    ""type"": ""record"",
    ""name"": ""Foo"",
    ""fields"": [      
    ]
}";

            var readerSchemaString = $@"{{
    ""type"": ""record"",
    ""name"": ""Foo"",
    ""fields"": [
        {{
            ""name"": ""x"",
            ""type"": {typeDefinition},
            ""default"": {defaultDefinition}
        }}
    ]
}}";
            var writerSchema = Schema.Parse(writerSchemaString);

            Stream stream;

            serialize(writerSchemaString,
                mkRecord(new object[] { }, (RecordSchema) writerSchema),
                out stream,
                out _);

            var output = deserialize<GenericRecord>(stream, writerSchema, Schema.Parse(readerSchemaString));

            Assert.AreEqual(expected, output.GetValue(0));
        }
```

You will note that some unit tests do not expect the way I think they should work, so those cases must be discussed. 

1. microsec logical types swallow/ignore the actual us values and work only for milliseconds.
2. decimal defaults do not work IMO as expected, however I might have been using them wrong in the test cases

Reply:
Thank you very much for the comment! I will add further tests. In regards to the two problematic cases you have discovered:

Reply:
I agree on the assessment of the the 2 issues. I would ignore the decimal unit test failures, however the micros are a  bigger issue. I would fix that first and put this PR on hold until that is merged.

Reply:
Sorry for the late reply - thank you very much for your PR!

Reply:
I have fixed the tests for decimals. Strangely, creating a new AvroDecimal for a TestCaseData instance leads to a stack overflow in NUnit - for this reason, the tests for decimal logical type are separate from the others.